### PR TITLE
GLDS: Text component 

### LINKED ISSUE #71415: Text component [build]: modify previous structure according to internal discussion
After some internal discussions in order to be sure of doing the best regarding the `Text` component, we will modify it by:

- [x] Revert the latest changes: `Text` as a unique component instead of `H1`, `H2`, `P` and so on
- [x] Rename `as` prop to `element`
- [x] Add an italic prop
- [x] Import `Text` component instead of the `TextElements` in the `unstable` package.

### PR DESCRIPTION
**Pull request by eledobleefe**
**What is this feature?**

After discussing internally, we have to modify the Text component in some aspects described in the issue. 

**Why do we need this feature?**

To improve the `Text` component and move it from the `unstable` package.

**Who is this feature for?**

Everybody

**Which issue(s) does this PR fix?**:

Fixes #71415 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by gillesdemey — APPROVED** [review]
Alerting changes LGTM ✅

File: packages/grafana-ui/src/components/Text/Text.tsx [hunk 1]
```diff
@@ -6,8 +6,8 @@ import { GrafanaTheme2, ThemeTypographyVariantTypes } from '@grafana/data';
 import { useStyles2 } from '../../themes';
 
 export interface TextProps {
-  /** Defines what HTML element is defined underneath */
-  as: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span' | 'p' | 'legend';
+  /** Defines what HTML element is defined underneath. "span" by default */
+  element?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span' | 'p' | 'legend';
```
Comment (mckn) [inline packages/grafana-ui/src/components/Text/Text.tsx:10]: I do like this change but I'm not sure it is worth introducing a breaking change just for the sake of changing the name of the property.

Is the Text component widely used by plugins? Have you done any research on this?

Comment (mckn) [reply packages/grafana-ui/src/components/Text/Text.tsx:10]: I saw (further down in the PR) that we expose this one via `unstable.ts`. I guess this is a new pattern we have introduced. @jackw something you have been involved in?

**Review by mckn — COMMENTED** [review]
LGTM! I added a couple of questions regarding the unstable.ts pattern.

**Review by L-M-K-B — COMMENTED** [review]
Looks good to me! 😃
There might be something wrong with the story but it can also be an issue on my computer. 😅

File: packages/grafana-ui/src/components/Text/Text.tsx (hunk 1 above)
Comment (eledobleefe) [reply packages/grafana-ui/src/components/Text/Text.tsx:10]: Hi @mckn,
The `unstable` package has been available only to grafana core.
We create this package to test the component and give the opportunity to use it by 'core' devs, so they can give us feedback regarding some concerns we had while we were building it.
With these latest changes, we are about to move the `Text` component to `grafana-ui`.

Comment (joshhunt) [reply packages/grafana-ui/src/components/Text/Text.tsx:10]: Considerable conversation here. https://raintank-corp.slack.com/archives/CT7S455SP/p1684324377579169

tl;dr - this component is not exposed publicly and was never stable a stable API. The point of this process is to make these "breaking changes" without actually impacting plugins.

File: packages/grafana-ui/src/components/Text/Text.internal.story.tsx [hunk 2]
```diff
@@ -49,83 +48,50 @@ const meta: Meta = {
...
+          but is comprised within the same block text
+        </Text>
       </StoryExample>
     </VerticalGroup>
   );
 };
 Example.parameters = {
   controls: {
-    exclude: ['variant', 'weight', 'textAlignment', 'truncate', 'color', 'children'],
+    exclude: ['variant', 'weight', 'textAlignment', 'truncate', 'italic', 'color', 'children'],
```
[the remark this answers is not in the document]
Comment (JoaoSilvaGrafana) [reply packages/grafana-ui/src/components/Text/Text.internal.story.tsx:67]: I think what's happening here is that the story doesn't receive `args` so it thinks it is not setup to use controls. I don't know what the common pattern is for how to write stories that shouldn't take any controls. I can go research on how that should work after this PR to fix it all over the place.

File: packages/grafana-ui/src/components/Text/Text.tsx [hunk 3]
```diff
@@ -16,22 +16,24 @@ export interface TextProps {
...
-        (theme) => getTextStyles(theme, variant, color, weight, truncate, textAlignment),
-        [color, textAlignment, truncate, weight, variant]
+        (theme) => getTextStyles(theme, variant, color, weight, truncate, italic, textAlignment),
+        [color, textAlignment, truncate, italic, weight, variant]
       )
     );
 
     return createElement(
-      as,
+      element,
```
Comment (JoaoSilvaGrafana) [inline packages/grafana-ui/src/components/Text/Text.tsx:36]: Do we need to readd here somewhere what is the default variant for each element? Feel like we have lost that in the process and it is now just going back to the default css?

File: public/app/features/alerting/unified/components/MetaText.tsx [hunk 4]
```diff
@@ -22,12 +22,12 @@ const MetaText = ({ children, icon, color = 'secondary', ...rest }: Props) => {
       // allow passing ARIA and data- attributes
       {...rest}
     >
-      <Span variant="bodySmall" color={color}>
+      <Text variant="bodySmall" color={color}>
```
Comment (JoaoSilvaGrafana) [inline public/app/features/alerting/unified/components/MetaText.tsx:25]: If we are keeping it like we had previously, `span` will be a `bodySmall` by default, so we can remove it from here and from any other places that use `span` with this variant 👍

File: packages/grafana-ui/src/components/Text/Text.internal.story.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (eledobleefe) [reply packages/grafana-ui/src/components/Text/Text.internal.story.tsx:67]: Thanks @JoaoSilvaGrafana! You're right. I've checked adding general `args` and now the `Example` excludes them.

Captura de pantalla 2023-07-14 a las 10 41 11

File: public/app/features/alerting/unified/components/MetaText.tsx (hunk 4 above)
Comment (eledobleefe) [reply public/app/features/alerting/unified/components/MetaText.tsx]: Should not the `span` element have a `body` variant? I think the variant `bodySmall` is used just in specific use cases so shouldn't we make the user select it when it is needed and not have it by default in `span` elements?

Comment (JoaoSilvaGrafana) [reply public/app/features/alerting/unified/components/MetaText.tsx]: Yeah I cant remember what the default should be, but I think we set it as `bodySmall` before, whichever is best though 😄

File: packages/grafana-ui/src/components/Text/utils.ts [hunk 5]
```diff
@@ -30,3 +30,32 @@ export const customColor = (color: TextProps['color'], theme: GrafanaTheme2): st
...
+  theme: GrafanaTheme2,
+  element: TextProps['element'],
+  variant?: keyof ThemeTypographyVariantTypes
+) => {
+  if (variant) {
+    return theme.typography[variant];
+  }
+  if (element === 'span') {
+    return;
+  }
```
Comment (JoaoSilvaGrafana) [inline packages/grafana-ui/src/components/Text/utils.ts:44]: we can put this inside the switch and return nothing just like here

File: packages/grafana-ui/src/components/Text/Text.internal.story.tsx [hunk 6]
```diff
@@ -39,93 +37,63 @@ const meta: Meta = {
...
-  variant: undefined,
-  weight: 'light',
-  textAlignment: 'center',
-  truncate: false,
-  color: 'primary',
-  children: 'This is a H1 component',
-};
-
-export const LegendComponent: StoryFn = (args) => {
+export const Basic: StoryFn = (args) => {
```
Comment (JoaoSilvaGrafana) [inline packages/grafana-ui/src/components/Text/Text.internal.story.tsx:83]: BTW I just tested this story with a `span` and `truncate=true` and it does not truncate, I think because `span` is by default `inline` and that doesn't truncate. Maybe the `truncate` should also set `display: block`? This will obviously break the text if it is surrounded by other text, but I can't think of why we would want to truncate that kind of text.

File: packages/grafana-ui/src/components/Text/Text.tsx (hunk 1 above)
Comment (mckn) [reply packages/grafana-ui/src/components/Text/Text.tsx]: Ah, ok - thanks for the clarification 🙏🏻

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx [hunk 7]
```diff
@@ -169,13 +169,7 @@ export function NestedFolderPicker({ value, onChange }: NestedFolderPickerProps)
         icon={value !== undefined ? 'folder' : undefined}
         ref={setTriggerRef}
       >
-        {selectedFolder.isLoading ? (
-          <Skeleton width={100} />
-        ) : (
-          <Text as="span" truncate>
-            {label ?? <Trans i18nKey="browse-dashboards.folder-picker.button-label">Select folder</Trans>}
```
[the remark this answers is not in the document]
Comment (eledobleefe) [reply public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:176]: Yes, I don't know the reason but I will fix it 👌

File: packages/grafana-ui/src/components/Text/utils.ts [hunk 8]
```diff
@@ -30,3 +30,31 @@ export const customColor = (color: TextProps['color'], theme: GrafanaTheme2): st
...
+  theme: GrafanaTheme2,
+  element: TextProps['element'],
+  variant?: keyof ThemeTypographyVariantTypes
+) => {
+  if (variant) {
+    return theme.typography[variant];
+  }
+  switch (element) {
+    case 'span':
+      return;
```
Comment (JoaoSilvaGrafana) [inline packages/grafana-ui/src/components/Text/utils.ts:44]: Let's leave a comment here explaining why span doesnt have a default as well, something like we want the default `Text` style to just take the parent's variant.

**Review by JoaoSilvaGrafana — APPROVED** [review]
👏 👏

File: public/app/core/components/NestedFolderPicker/NestedFolderList.tsx [hunk 9]
```diff
@@ -209,9 +209,8 @@ function Row({ index, style: virtualStyles, data }: RowProps) {
         )}
 
         <label className={styles.label} id={labelId}>
-          <Text as="span" truncate>
-            {item.title}
-          </Text>
+          {/* TODO: text is not truncated properly, it still overflows the container */}
```
Comment (ashharrison90) [inline public/app/core/components/NestedFolderPicker/NestedFolderList.tsx:212]: @eledobleefe friendly reminder to double check your merges so we aren't readding comments that don't apply anymore 🙏

Comment (eledobleefe) [reply public/app/core/components/NestedFolderPicker/NestedFolderList.tsx:212]: Hi @ashharrison90,
I wrongly linked it to the [Tooltip task](https://github.com/grafana/grafana/pull/68578) regarding the `truncate` prop. Do you want me to delete this comment in that PR? Or should I do it in a new one now?

Comment (ashharrison90) [reply public/app/core/components/NestedFolderPicker/NestedFolderList.tsx:212]: i fixed it [here](https://github.com/grafana/grafana/pull/72186), more just an fyi than anything 👍
